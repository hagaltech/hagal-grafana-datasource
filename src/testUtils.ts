import {of, throwError} from 'rxjs';
import {DataSourceInstanceSettings} from '@grafana/data';
import _ from 'lodash';

import {DataSource, PROXY_PATH} from './datasource';
import {DataSourceConnectionOptions, ProxyRequestDataOptions} from './types';

import timeseriesFilterData from 'mock/timeseriesFilterData.json';

type MockData = {
  [key: string]: any;
};
type ErrorObject = {
  status?: number;
  data?: {
    error: {
      code: number;
      message: string;
    };
  };
};
type mockErrorFunctionType = ((data: ProxyRequestDataOptions) => ErrorObject) | undefined;

export function getItemsResponseObject(items: MockData) {
  return {
    data: {
      items,
    },
    status: 200
  };
}

export function getTimeseriesResponse(data: ProxyRequestDataOptions, dpNumber = 5) {
  const {aggregates, items} = data;
  const aggregate = aggregates ? aggregates[0] : 'value';

  const datapoints = new Array(dpNumber).fill(null).map((_, i) => ({
    timestamp: i * 600000 + 1549336675000,
    [aggregate]: i,
  }));

  return _.map(items, ({ id }) => ({
    id,
    datapoints,
  }));
}

const instanceSettings: DataSourceInstanceSettings<DataSourceConnectionOptions> = {
  id: 1,
  type: 'hagal-api-datasource',
  name: 'Hagal API Data',
  access: 'direct'
} as DataSourceInstanceSettings<DataSourceConnectionOptions>;

export const getMockedDataSource = () => new DataSource(instanceSettings);

const mockData: MockData = {
  [`/${PROXY_PATH}/timeseries/filter`]: () => timeseriesFilterData,
  [`/${PROXY_PATH}/datapoints/aggregate`]: getTimeseriesResponse,
  [`/${PROXY_PATH}/datapoints/list`]: getTimeseriesResponse,
};

export function getMockedFetch(mockErrorFunction?: mockErrorFunctionType) {
  return ({ url, data }: { url: string, data: ProxyRequestDataOptions }) => {
    if (mockErrorFunction) {
      const error = mockErrorFunction(data);
      if (error) {
        return throwError(() => error);
      }
    }

    if (mockData[url]) {
      return of(getItemsResponseObject(mockData[url](data)));
    }

    return of(getItemsResponseObject([]));
  };
}
