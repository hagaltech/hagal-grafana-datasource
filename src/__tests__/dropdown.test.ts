import {getMockedDataSource, getMockedFetch} from 'testUtils';
import {SelectableValue} from '@grafana/data';

const mockedFetch = jest.fn(getMockedFetch());
jest.mock('@grafana/runtime', () => ({
  getBackendSrv: jest.fn(() => ({
    fetch: mockedFetch
  }))
}));

type Mock = jest.Mock;

const ds = getMockedDataSource();

describe('Timeseries Dropdown Options', () => {
  describe('Given a request for timeseries options', () => {
    let result: Array<SelectableValue<string>>;

    beforeAll(async () => {
      result = await ds.getOptionsForDropdown('');
    });

    it('should generate the correct request', () => {
      expect(mockedFetch).toHaveBeenCalledTimes(1);
      expect((mockedFetch as Mock).mock.calls[0][0]).toMatchSnapshot();
    });

    it('should return all timeseries', () => {
      expect(result).toMatchSnapshot();
    });
  })

  describe('Given a search string for timeseries options', () => {
    let result: Array<SelectableValue<string>>;

    beforeAll(async () => {
      result = await ds.getOptionsForDropdown('SOC');
    });

    it('should generate the correct request', () => {
      expect(mockedFetch).toHaveBeenCalledTimes(1);
      expect((mockedFetch as Mock).mock.calls[0][0]).toMatchSnapshot();
    });

    it('should return timeseries with SOC in their id or label', () => {
      expect(result).toMatchSnapshot();
    });
  })
})
