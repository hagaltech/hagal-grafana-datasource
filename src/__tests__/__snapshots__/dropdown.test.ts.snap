// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeseries Dropdown Options Given a request for timeseries options should generate the correct request 1`] = `
{
  "data": {},
  "method": "POST",
  "url": "/api/timeseries/filter",
}
`;

exports[`Timeseries Dropdown Options Given a request for timeseries options should return all timeseries 1`] = `
[
  {
    "id": "bms-BatteryStringCurrent",
    "label": "Battery String Current Råde",
    "value": "bms-BatteryStringCurrent",
  },
  {
    "id": "bms-BatteryStringSelfCheckAlarmCode",
    "label": "Battery String Self Check Alarm Code Råde",
    "value": "bms-BatteryStringSelfCheckAlarmCode",
  },
  {
    "id": "bms-BatteryStringVoltage",
    "label": "Battery String Voltage Råde",
    "value": "bms-BatteryStringVoltage",
  },
  {
    "id": "bms-BatteryStringSOH",
    "label": "Battery String SOH Råde",
    "value": "bms-BatteryStringSOH",
  },
  {
    "id": "bms-BatteryStringSOC",
    "label": "Battery String SOC Råde",
    "value": "bms-BatteryStringSOC",
  },
  {
    "id": "bms-BatteryStringAlarmCodeLevel3",
    "label": "Battery String Alarm Code Level3 Råde",
    "value": "bms-BatteryStringAlarmCodeLevel3",
  },
  {
    "id": "bms-BatteryStringAlarmCodeLevel1",
    "label": "Battery String Alarm Code Level1 Råde",
    "value": "bms-BatteryStringAlarmCodeLevel1",
  },
  {
    "id": "bms-BatteryClusterEnergyFlowState",
    "label": "Battery Cluster Energy Flow State Råde",
    "value": "bms-BatteryClusterEnergyFlowState",
  },
  {
    "id": "bms-BatteryStringAlarmCodeLevel2",
    "label": "Battery String Alarm Code Level2 Råde",
    "value": "bms-BatteryStringAlarmCodeLevel2",
  },
]
`;

exports[`Timeseries Dropdown Options Given a search string for timeseries options should generate the correct request 1`] = `
{
  "data": {},
  "method": "POST",
  "url": "/api/timeseries/filter",
}
`;

exports[`Timeseries Dropdown Options Given a search string for timeseries options should return timeseries with SOC in their id or label 1`] = `
[
  {
    "id": "bms-BatteryStringSOC",
    "label": "Battery String SOC Råde",
    "value": "bms-BatteryStringSOC",
  },
]
`;
